File: cli/src/specsfy_cli/catalog.py
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .github import api_headers


DEFAULT_CATALOG_URL = (
    "https://api.github.com/repos/promovaweb/specsfy/"
    "contents/specialists/catalog.json?ref=main"
)


@dataclass(frozen=True)
class CatalogEntry:
    name: str
    description: str
    category: str
    tags: tuple[str, ...]
    files: tuple[str, ...]
    dependencies: tuple[str, ...]
    requires: tuple[str, ...] = ()

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> "CatalogEntry":
        detect = value.get("detect", {})
        return cls(
            name=value["name"],
            description=value["description"],
            category=value["category"],
            tags=tuple(value.get("tags", [])),
            files=tuple(detect.get("files", [])),
            dependencies=tuple(detect.get("dependencies", [])),
            requires=tuple(value.get("requires", [])),
        )


class Catalog:
    def __init__(self, entries: list[CatalogEntry]) -> None:
        self.entries = sorted(entries, key=lambda entry: entry.name)
        self._by_name = {entry.name: entry for entry in entries}

    @classmethod
    def from_payload(cls, payload: dict[str, Any]) -> "Catalog":
        if payload.get("schema_version") != 1:
            raise ValueError("versão de catálogo não suportada")
        return cls([CatalogEntry.from_dict(value) for value in payload["skills"]])

    @classmethod
    def from_path(cls, path: Path) -> "Catalog":
        return cls.from_payload(json.loads(path.read_text(encoding="utf-8")))

    @classmethod
    def fetch(cls, url: str | None = None) -> "Catalog":
        override = os.environ.get("SPECSFY_SPECIALISTS_CATALOG")
        if override:
            return cls.from_path(Path(override).expanduser().resolve())
        request = urllib.request.Request(
            url or DEFAULT_CATALOG_URL,
            headers=api_headers(
                "specsfy-cli",
                accept="application/vnd.github.raw+json",
            ),
        )
        try:
            with urllib.request.urlopen(request, timeout=20) as response:
                return cls.from_payload(json.load(response))
        except urllib.error.HTTPError as error:
            if error.code in {401, 403, 404}:
                raise RuntimeError(
                    "catálogo de especialistas indisponível; autentique o GitHub "
                    "com `gh auth login` ou defina GH_TOKEN"
                ) from error
            raise

    def require(self, name: str) -> CatalogEntry:
        if not name.startswith("specsfy-specialist-"):
            raise ValueError("a skill especialista deve usar o prefixo specsfy-specialist-")
        try:
            return self._by_name[name]
        except KeyError as error:
            raise ValueError(f"skill não encontrada no catálogo: {name}") from error

    def resolve(self, names: list[str]) -> list[CatalogEntry]:
        resolved: list[CatalogEntry] = []
        completed: set[str] = set()
        active: set[str] = set()

        def visit(name: str) -> None:
            if name in completed:
                return
            if name in active:
                raise ValueError(
                    f"dependência circular entre especialistas: {name}"
                )
            entry = self.require(name)
            active.add(name)
            for required_name in entry.requires:
                visit(required_name)
            active.remove(name)
            completed.add(name)
            resolved.append(entry)

        for name in names:
            visit(name)
        return resolved

    def detect(self, project: Path) -> list[CatalogEntry]:
        dependencies = _read_dependencies(project)
        detected = []
        for entry in self.entries:
            file_match = any((project / marker).exists() for marker in entry.files)
            dependency_match = any(
                dependency in dependencies for dependency in entry.dependencies
            )
            if file_match or dependency_match:
                detected.append(entry)
        return detected


def _read_dependencies(project: Path) -> set[str]:
    dependencies: set[str] = set()
    for filename in ("package.json", "composer.json"):
        path = project / filename
        if not path.is_file():
            continue
        try:
            payload = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            continue
        for key in (
            "dependencies",
            "devDependencies",
            "require",
            "require-dev",
            "peerDependencies",
        ):
            value = payload.get(key, {})
            if isinstance(value, dict):
                dependencies.update(value)
    return dependencies

File: cli/src/catalog.ts
/**
 * Catálogo de especialistas e detecção de tecnologias do projeto.
 */

import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { apiHeaders } from "./github.js";
import { isFile, pathExists, resolvePath } from "./filesystem.js";

export const DEFAULT_CATALOG_URL =
  "https://api.github.com/repos/promovaweb/specsfy/contents/specialists/catalog.json?ref=main";

/** Entrada normalizada do catálogo de especialistas. */
export interface CatalogEntryValue {
  name: string;
  description: string;
  category: string;
  tags: string[];
  files: string[];
  dependencies: string[];
  requires: string[];
}

interface CatalogPayload {
  schema_version: number;
  skills: Array<{
    name: string;
    description: string;
    category: string;
    tags?: string[];
    requires?: string[];
    detect?: { files?: string[]; dependencies?: string[] };
  }>;
}

/** Catálogo validado com resolução de dependências entre especialistas. */
export class Catalog {
  readonly entries: CatalogEntryValue[];
  readonly #byName: Map<string, CatalogEntryValue>;

  constructor(entries: CatalogEntryValue[]) {
    this.entries = [...entries].sort((left, right) =>
      left.name.localeCompare(right.name),
    );
    this.#byName = new Map(entries.map((entry) => [entry.name, entry]));
  }

  /** Constrói o catálogo a partir do schema JSON público. */
  static fromPayload(payload: unknown): Catalog {
    if (
      !payload ||
      typeof payload !== "object" ||
      (payload as Partial<CatalogPayload>).schema_version !== 1 ||
      !Array.isArray((payload as Partial<CatalogPayload>).skills)
    ) {
      throw new Error("versão de catálogo não suportada");
    }
    const source = payload as CatalogPayload;
    return new Catalog(
      source.skills.map((value) => ({
        name: value.name,
        description: value.description,
        category: value.category,
        tags: value.tags ?? [],
        files: value.detect?.files ?? [],
        dependencies: value.detect?.dependencies ?? [],
        requires: value.requires ?? [],
      })),
    );
  }

  /** Lê o catálogo de um arquivo local. */
  static async fromPath(path: string): Promise<Catalog> {
    return Catalog.fromPayload(JSON.parse(await readFile(resolvePath(path), "utf8")));
  }

  /** Obtém o catálogo remoto ou usa o override de ambiente. */
  static async fetch(url = DEFAULT_CATALOG_URL): Promise<Catalog> {
    const override = process.env.SPECSFY_SPECIALISTS_CATALOG;
    if (override) return Catalog.fromPath(override);
    const response = await fetch(url, {
      headers: await apiHeaders(
        "specsfy-cli",
        "application/vnd.github.raw+json",
      ),
      signal: AbortSignal.timeout(20_000),
    });
    if (!response.ok) {
      if ([401, 403, 404].includes(response.status)) {
        throw new Error(
          "catálogo de especialistas indisponível; autentique o GitHub " +
            "com `gh auth login` ou defina GH_TOKEN",
        );
      }
      throw new Error(`falha ao consultar catálogo: HTTP ${response.status}`);
    }
    return Catalog.fromPayload((await response.json()) as unknown);
  }

  /** Exige uma skill existente com o namespace de especialistas. */
  require(name: string): CatalogEntryValue {
    if (!name.startsWith("specsfy-specialist-")) {
      throw new Error(
        "a skill especialista deve usar o prefixo specsfy-specialist-",
      );
    }
    const entry = this.#byName.get(name);
    if (!entry) throw new Error(`skill não encontrada no catálogo: ${name}`);
    return entry;
  }

  /** Resolve requisitos em ordem topológica e rejeita ciclos. */
  resolve(names: string[]): CatalogEntryValue[] {
    const resolved: CatalogEntryValue[] = [];
    const completed = new Set<string>();
    const active = new Set<string>();
    const visit = (name: string): void => {
      if (completed.has(name)) return;
      if (active.has(name)) {
        throw new Error(`dependência circular entre especialistas: ${name}`);
      }
      const entry = this.require(name);
      active.add(name);
      for (const required of entry.requires) visit(required);
      active.delete(name);
      completed.add(name);
      resolved.push(entry);
    };
    for (const name of names) visit(name);
    return resolved;
  }

  /** Detecta entradas por arquivos marcadores e dependências declaradas. */
  async detect(project: string): Promise<CatalogEntryValue[]> {
    const root = resolvePath(project);
    const dependencies = await readDependencies(root);
    const detected: CatalogEntryValue[] = [];
    for (const entry of this.entries) {
      const matches = await Promise.all(
        entry.files.map((marker) => pathExists(join(root, marker))),
      );
      if (
        matches.some(Boolean) ||
        entry.dependencies.some((dependency) => dependencies.has(dependency))
      ) {
        detected.push(entry);
      }
    }
    return detected;
  }
}

async function readDependencies(project: string): Promise<Set<string>> {
  const dependencies = new Set<string>();
  for (const filename of ["package.json", "composer.json"]) {
    const path = join(project, filename);
    if (!(await isFile(path))) continue;
    let payload: Record<string, unknown>;
    try {
      payload = JSON.parse(await readFile(path, "utf8")) as Record<string, unknown>;
    } catch {
      continue;
    }
    for (const key of [
      "dependencies",
      "devDependencies",
      "require",
      "require-dev",
      "peerDependencies",
    ]) {
      const value = payload[key];
      if (value && typeof value === "object" && !Array.isArray(value)) {
        for (const dependency of Object.keys(value)) dependencies.add(dependency);
      }
    }
  }
  return dependencies;
}
